Split show hide email checkbox, added edit profile page load for issue #764

rbillings opened the pull request:
For issue #764

bobsilverberg commented:
Thanks for taking this on, @rbillings. I have a few comments which I will add to the diff.

bobsilverberg commented on pages/desktop/base.py:
```diff
@@ -248,6 +248,7 @@ def click_edit_profile(self):
                 click().perform()
 
             from pages.desktop.user import EditProfile
+            WebDriverWait(self.selenium, self.timeout).until(lambda s: self.selenium.title)
```
If the intent here is to wait for the page to be fully loaded, this isn't necessarily going to do it. This is only going to wait until selenium sees a valid title for the page, which might still be too soon. I suggest removing this code, and adding the wait into the methods that interact with the checkbox.

In issue #764 @davehunt suggests waiting for the checkbox to be displayed and to have the correct state, so that code really belongs where we are doing the click. See my comments below.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -119,8 +119,15 @@ def click_update_account(self):
-    def change_hide_email_state(self):
-        self.selenium.find_element(*self._hide_email_checkbox).click()
+    def hide_email_address(self):
+        email_checkbox = self.selenium.find_element(*self._email_checkbox)
```
It is in this method, as well as `show_email_address` that it would make sense to add the wait for the checkbox to be displayed and in the expected state.

rbillings replied:
What you are suggesting for line 128 displays on line 129- what is it that is missing?

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -119,8 +119,15 @@ def click_update_account(self):
-        self.selenium.find_element(*self._hide_email_checkbox).click()
+    def hide_email_address(self):
+        email_checkbox = self.selenium.find_element(*self._email_checkbox)
+        if not email_checkbox.is_selected():
```
Instead of using an `if` here, change it to a wait so that we actually wait for the checkbox to be not selected. If that wait times out then the test will fail which makes sense, as the state we expected was not met.

rbillings replied:
I added the wait above which verifies that the checkbox displays, this particular "if" determines whether or not the checkbox is already selected. I don't understand why another wait would be appropriate as the status of whether or not the checkbox is selected doesn't change with time. It does make sense to add a wait after line 125 as the state should then match the expected state.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -119,8 +119,15 @@ def click_update_account(self):
+    def hide_email_address(self):
+        email_checkbox = self.selenium.find_element(*self._email_checkbox)
+        if not email_checkbox.is_selected():
+                self.selenium.find_element(*self._email_checkbox).click()
```
After clicking on the element, you could then wait for the new state of the checkbox to be what we now expect it to be, again ensuring that the test does not proceed unless the page is in the state that we expect.

rbillings replied:
So I added a wait for the checkbox upon loading edit profile, another wait on hide_email_address to make sure it's visible- do we really need a third wait to ensure that the checkbox was checked? That will be two wait statements in one method- three if I also add one for line 124.

bobsilverberg commented:
Good work @rbillings! Let me know if you have any questions on any of my comments.

rbillings commented:
I removed the wait from base, added a specific checkbox wait and a verification that the checkbox ends up in the correct expected state- let me know if that does it.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -119,8 +120,19 @@ def click_update_account(self):
+        email_checkbox = self.selenium.find_element(*self._email_checkbox)
+        WebDriverWait(self.selenium, self.timeout).until(expected_conditions.visibility_of(email_checkbox));
+        if not email_checkbox.is_selected():
+                self.selenium.find_element(*self._email_checkbox).click()
```
You have a good point about the extra wait I suggested, and the fact that the state isn't likely to change over time, so an `if` is appropriate, but I think you need to fail the test if the `if` evaluates to indicate that the expected state is not present. For example, in this current code, if the checkbox is currently selected, rather than fail the test you simply do not click on the checkbox. I suggest that is the checkbox is currently selected, but we expected it not to be selected, then we need to fail the test.

The `return email_checkbox.is_selected() == True` that you add below doesn't actually do anything, as you are not checking the return value from this method (nor would I expect you to). I am guessing that this `return` was an attempt to notify the test if the page is not in the expected state after clicking the checkbox, but it will not do that.

What would do that would be to replace this `return` with a wait that does wait for the checkbox to be in the correct state. If you make those suggested changes then the following should be in place:

1. If the checkbox is not in the expected state before clicking - fail the test.
2. If the checkbox is not in the expected after before clicking - fail the test (via a wait timeout).

rbillings replied:
The action of the method is to hide the email via the checkbox. If I do what you suggest then it will fail the test if the email is already hidden, which doesn't seem right to me - because it isn't failing to hide the email, it's failing to be unhidden before the test started. If it fails before the test starts then the failure should be somewhere else.

bobsilverberg replied:
The point is that you should have an expectation of the state before the test runs. If you expect that the email will already be hidden, then you shouldn't be calling `hide_email_address` in the test. A good test should always have a known state at the beginning of the test, and a known state at any point. Are you saying that we cannot know the initial state? That sometimes it will start hidden and sometimes it will not? If that is the case then I suggest that there is a problem with the test.

bobsilverberg commented:
@rbillings I added a new comment to the diff about your most recent changes.

rbillings commented:
This is ready for review.

rbillings commented:
@bobsilverberg r+?

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -112,6 +113,10 @@ def details_header_text(self):
         return self.selenium.find_element(*self._details_locator).text
 
     @property
+    def checkbox_status(self):
```
This name seems to vague for what the property is. This returns `True` if the checkbox is checked and `False` if it is not, but I wouldn't intuit that from the name `checkbox_status`. Perhaps something like `is_hide_email_checked` would be more descriptive.

Having said that, I don't think we actually need this method at all as the test doesn't really need it if the page object handles `hide_email_address` appropriately.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -89,7 +90,7 @@ class EditProfile(Base):
     _details_locator = (By.CSS_SELECTOR, "#profile-detail > legend")
     _notification_locator = (By.CSS_SELECTOR, "#acct-notify > legend")
-    _hide_email_checkbox = (By.ID, 'id_emailhidden')
+    _email_checkbox = (By.ID, 'id_emailhidden')
```
The variable should have the `_locator` suffix as do all the others. Also, I actually prefer the old name. This checkbox is checked when the user wants to hide their email address, correct? If that's the case then just `email_checkbox` is less accurate than `hide_email_checkbox`.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -119,8 +124,28 @@ def click_update_account(self):
-        self.selenium.find_element(*self._hide_email_checkbox).click()
+    def hide_email_address(self):
+        email_checkbox = self.selenium.find_element(*self._email_checkbox)
+        WebDriverWait(self.selenium, self.timeout).until(expected_conditions.visibility_of(email_checkbox));
```
We do not put a semicolon at the end of Python statements.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -119,8 +124,28 @@ def click_update_account(self):
+
+        if email_checkbox.is_selected():
+            err_msg = 'Email is already hidden'
+            return err_msg
```
You cannot return an error message here as nothing will be done with it. If you look at the calling code you wrote (https://github.com/mozilla/Addon-Tests/pull/765/files#diff-2e5bbbef9e51c633c062636e9f31733fR94), it just calls the method but doesn't do anything with the return value. So even if this code was encountered, and you returned `err_msg`, nothing would happen - the test would just continue on its merry way.

What you need to do is throw an exception here, which will stop the test immediately, as that's really what we want to happen if our expectations are not met. You could use:

`raise Exception('Email is already hidden')`

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -119,8 +124,28 @@ def click_update_account(self):
+            err_msg = 'Email is already hidden'
+            return err_msg
+
+        if not email_checkbox.is_selected():
```
This `if` statement is not needed as the code will never reach this point unless the condition is true. I.e., if the email_checkbox was selected we would have already thrown the above error (or returned the error message, in your current code).

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -119,8 +124,28 @@ def click_update_account(self):
+
+        if not email_checkbox.is_selected():
+            self.selenium.find_element(*self._email_checkbox).click()
+        return email_checkbox.is_selected() == True
```
Again you are returning something from a method that isn't expected to return anything, so this code won't do anything. I think what we need to do here is instead to wait for the email checkbox to be checked. You can do this as above with `WebDriverWait`, e.g., `WebDriverWait(self.selenium, self.timeout).until(lambda s: email_checkbox.is_selected())`

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -119,8 +124,28 @@ def click_update_account(self):
+            self.selenium.find_element(*self._email_checkbox).click()
+        return email_checkbox.is_selected() == True
+
+    def show_email_address(self):
```
The same changes as above need to be applied to this method to, i.e.,
1. Find the checkbox element.
2. Wait for it to be visible.
3. If it is already unchecked, raise an exception.
4. Click it (which should uncheck it).
5. Wait for it to be unchecked.

bobsilverberg commented on tests/desktop/test_users_account.py:
```diff
@@ -86,13 +86,21 @@ def test_hide_email_checkbox_works(self, mozwebqa):
         edit_profile_page = home_page.header.click_edit_profile()
-        edit_profile_page.change_hide_email_state()
-        edit_profile_page.click_update_account()
+        # make sure email checkbox isn't already selected
```
With the above changes made to the page object, you shouldn't need any of this code. If we expect the box to be initially unchecked (and I think that's true, but I could be wrong), then all you need to do is call `edit_profile_page.hide_email_address()` and it will take care of everything for you.

bobsilverberg commented on tests/desktop/test_users_account.py:
```diff
@@ -86,13 +86,21 @@ def test_hide_email_checkbox_works(self, mozwebqa):
         final_state = view_profile_page.is_email_field_present
 
-        Assert.not_equal(initial_state, final_state, 'The initial and final states are the same. The profile change failed.')
+        Assert.equal(view_profile_page.is_email_field_present, False, 'The email address checkbox is not hidden')
```
I think the failure here is "The email address is not hidden", not "The email address *checkbox* is not hidden".

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -117,10 +118,31 @@ def notification_header_text(self):
+
+        if email_checkbox.is_selected():
+            raise Exception('Email is already hidden')
+        else:
```
You don't need this `else:` because if the email_checkbox _is_ already selected then the `raise` on line 130 will cause this method to exit, so the next lines will not execute. This means you can remove this `else:` and just put the call to `click()` on the next line.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -117,10 +118,31 @@ def notification_header_text(self):
-
-    def change_hide_email_state(self):
-        self.selenium.find_element(*self._hide_email_checkbox).click()
+        WebDriverWait(self.selenium, self.timeout).until
```
I'm not a big fan of the way this is indented (or rather, not indented). I'm actually surprised that this isn't a `pep8` violation, but it is not. I think it's more readable (and more consistent with how we usually write our code) to do it like:

```
        WebDriverWait(self.selenium, self.timeout).until(
            expected_conditions.visibility_of(email_checkbox))
```

The same goes for the other calls to `WebDriverWait` that are lacking this indentation on the 2nd line.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -117,10 +118,31 @@ def notification_header_text(self):
+        WebDriverWait(self.selenium, self.timeout).until
+        (lambda s: email_checkbox.is_selected())
+
+    def show_email_address(self):
```
This method is not currently being used, so in that case should be included. Having said that, I think it needs to be used (I'll comment more on this later), so we can keep it.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -117,10 +118,31 @@ def notification_header_text(self):
+            raise Exception('Email is already hidden')
+        else:
+            self.selenium.find_element(*self._hide_email_checkbox_locator).click()
+        WebDriverWait(self.selenium, self.timeout).until
```
We probably don't need this wait. We don't generally wait for UI elements to change that just represent clicking on a form field.

bobsilverberg commented on pages/desktop/user.py:
```diff
@@ -117,10 +118,31 @@ def notification_header_text(self):
+            raise Exception('Email is already shown')
+        else:
+            self.selenium.find_element(*self._hide_email_checkbox_locator).click()
+        return email_checkbox.is_selected() == False
```
We don't need to return anything here. Please remove this line.

bobsilverberg commented on tests/desktop/test_users_account.py:
```diff
@@ -75,7 +75,7 @@ def test_user_can_access_the_view_profile_page(self, mozwebqa):
     @pytest.mark.native
     @pytest.mark.login
-    def test_hide_email_checkbox_works(self, mozwebqa):
+    def test_hide_email_checkbox(self, mozwebqa):
```
There is a merge conflict with this file, and you'll find that the issue is that this line needs to change to:

`def test_hide_email_checkbox(self, mozwebqa, existing_user):`

bobsilverberg commented on tests/desktop/test_users_account.py:
```diff
@@ -86,13 +86,12 @@ def test_hide_email_checkbox_works(self, mozwebqa):
         initial_state = view_profile_page.is_email_field_present
```
The variable is never used, so it should be removed.

bobsilverberg commented on tests/desktop/test_users_account.py:
```diff
@@ -86,13 +86,12 @@ def test_hide_email_checkbox_works(self, mozwebqa):
         edit_profile_page.click_update_account()
-
         view_profile_page = home_page.header.click_view_profile()
         final_state = view_profile_page.is_email_field_present
```
This variable is also never used, so should be removed.

bobsilverberg commented on tests/desktop/test_users_account.py:
```diff
@@ -86,13 +86,12 @@ def test_hide_email_checkbox_works(self, mozwebqa):
         final_state = view_profile_page.is_email_field_present
 
-        Assert.not_equal(initial_state, final_state, 'The initial and final states are the same. The profile change failed.')
+        Assert.equal(view_profile_page.is_email_field_present, False, 'The email address is not hidden')
```
There is a problem with the test. If you run it twice it will fail (at least it does for me). The first time it clicks the box, and then the second time it will find the box already selected. Perhaps the test should both hide and then reshow the email address. First it could hide it, by checking the box, and then verify that the email address is hidden, and then it could show it again, by unchecking the box (which is where you'd make use of `show_email_address`) and then verify that it appears again.

There is still an issue that we are using an existing user and therefore cannot be sure of the initial state of the checkbox. If all goes well with our test, it should be in an unchecked state at the end of the test, but if the test fails or errors partway through then it might leave the user in a state where the checkbox is checked, in which case the next run of the test will fail.

I don't suppose we have the option of using a new user for each run of the test?

Another option would be to have a setup step where we make sure that the box is unchecked, and if it is not then we uncheck it.

Do you follow what I'm saying above? What do you think?

bobsilverberg commented:
@rbillings The latest version is really good. I have added a few comments to be addressed and I raised a question for discussion. Also, there is a merge conflict that needs to be resolved (I mentioned the solution in a comment). I am sorry it has taken so long for me to re-review this, but it fell off my radar. In future it would be helpful to me if you would ping me in the pull request to let me know it is ready for another review. I don't get informed when new commits are added, unless I am pinged in a PR comment.

rbillings commented:
@bobsilverberg- this is why the original test determined if the checkbox changed status, not whether it was shown or hidden. If it finds the email already hidden it fails correctly. To make it stop, then go run the show email address and then re-do the hide email address sounds wrong. Also, there are 3 static accounts used so it cannot be a new user each time. 

The only solution I can see from what you've suggested is to have the pre-set-up fix, but honestly I'd prefer to go back to just changing the status of the checkbox as this has become extremely complicated for a checkbox test. If you feel strongly that a hide email test is impt/valid then I will correct it as you suggest.

davehunt commented:
> This is why the original test determined if the checkbox changed status, not whether it was shown or hidden. If it finds the email already hidden it fails correctly. To make it stop, then go run the show email address and then re-do the hide email address sounds wrong.

Unfortunately this means the test was non-deterministic. Essentially it tested different functionality each time it ran - first it tested that the email could be shown, then hidden, etc. By setting an expected initial state we can test the same functionality each time.

> Also, there are 3 static accounts used so it cannot be a new user each time. 

As long as we're using an existing account, this test could become out of sync. I dont think there's any way currently of generating a new user for AMO tests, so I suspect we're stuck with this for now.

> The only solution I can see from what you've suggested is to have the pre-set-up fix, but honestly I'd prefer to go back to just changing the status of the checkbox as this has become extremely complicated for a checkbox test. If you feel strongly that a hide email test is impt/valid then I will correct it as you suggest.

We may need to compromise. While I agree with all of @bobsilverberg's suggestions, if we can't guarantee we're using a user with a known state the test could get out of sync, which would need manual intervention. Perhaps we can comment in the test that we're at least aware that it's not ideal.

I would suggest changing the test to:

  1. Determine the email visibility
  1. Toggle the checkbox
  1. Check the email visibility has changed 
  1. Toggle the checkbox
  1. Check the email visibility has changed

This way we'll be testing both show and hide every time the test runs, regardless of the initial state. Once we can create new users in our suite then I would recommend revisiting tests like this to set an expected initial state.

davehunt commented on pages/desktop/user.py:
```diff
@@ -117,10 +119,27 @@ def notification_header_text(self):
+            expected_conditions.visibility_of(email_checkbox))
+
+        if email_checkbox.is_selected():
+            raise Exception('Email is already hidden')
```
Note that my earlier comment may mean this code will change. This is more of a recommendation of best practice. Rather than raising an exception, we can do an assert here:

```python
asset not email_checkbox.is_selected(), 'Email is already hidden'
```

davehunt commented on pages/desktop/user.py:
```diff
@@ -117,10 +119,27 @@ def notification_header_text(self):
+
+        if email_checkbox.is_selected():
+            raise Exception('Email is already hidden')
+            self.selenium.find_element(*self._hide_email_checkbox_locator).click()
```
Note that this code would never be reached, as it's within an `if` statement that raises an exception immediately before.

davehunt replied:
Also note that we have already found this element, so there's no need to repeat `find_element`.

davehunt commented on tests/desktop/test_users_account.py:
```diff
@@ -83,16 +83,13 @@ def test_hide_email_checkbox_works(self, mozwebqa):
-        final_state = view_profile_page.is_email_field_present
 
-        Assert.not_equal(initial_state, final_state, 'The initial and final states are the same. The profile change failed.')
+        Assert.equal(view_profile_page.is_email_field_present, False, 'The email address is not hidden')
```
If we're just asserting if something is True or False we don't need to use `Assert.equal`. We should either use `assert condition`, `assert not condition`, `Assert.true(condition)` or `Assert.false(condition)`. Preference to `assert ...` as we're moving away from unittestzero.

rbillings commented:
I was thinking about this last night. I agree that the original test is not ideal, as is the fact that we have static users. I believe there is a PII case for ensuring the email address IS hidden. @davehunt - your suggestion would ensure that the checkbox works and the state changes, would you also agree that it specifically confirms that email addresses are hidden?

davehunt commented:
> I was thinking about this last night. I agree that the original test is not ideal, as is the fact that we have static users. I believe there is a PII case for ensuring the email address IS hidden. @davehunt - your suggestion would ensure that the checkbox works and the state changes, would you also agree that it specifically confirms that email addresses are hidden?

Yes, my proposal would check both hidden and shown, we just won't care which order they occur in. It's not ideal, but possibly the best we can do without the ability to generate new users and therefore have a known initial state.

rbillings commented:
The original test found the checkbox and clicked it, and verified the state changed. This would essentially be clicking it twice to make sure both states are covered. @bobsilverberg had strong opinions about why this wasn't a good idea so I'm going to let him chime in before proceeding with this.

bobsilverberg commented:
@rbillings @davehunt I was originally concerned about a page object method that just toggles the check box, regardless of which state it is already in, and suggested changing it so that we have two methods; one that shows the email, and one that hides the email. Each of those methods would fail if the expected pre-condition were not true. I.e., if you call "show email" but the email is already shown a failure would occur. I still feel like that is the best way to write the page object, but the problem now is that we don't know which of those methods to call first, as we don't know the current state of the user. 

@davehunt's suggested test flow above addresses that issue by checking the state of the user at the beginning of the test. I think the only thing in question to me is whether we keep two separate methods which either show or hide the email address, or whether we just use a toggle method which simply clicks the check box to change the state regardless of what the current state is. Either will work, and I'm fine with either one.

@davehunt Which do you suggest?

davehunt commented:
> @rbillings @davehunt I was originally concerned about a page object method that just toggles the check box, regardless of which state it is already in, and suggested changing it so that we have two methods; one that shows the email, and one that hides the email. Each of those methods would fail if the expected pre-condition were not true. I.e., if you call "show email" but the email is already shown a failure would occur. I still feel like that is the best way to write the page object, but the problem now is that we don't know which of those methods to call first, as we don't know the current state of the user. 

I agree this would be ideal, however if we can't have an expected initial value for the checkbox then we won't be able to use these methods.

> @davehunt's suggested test flow above addresses that issue by checking the state of the user at the beginning of the test. I think the only thing in question to me is whether we keep two separate methods which either show or hide the email address, or whether we just use a toggle method which simply clicks the check box to change the state regardless of what the current state is. Either will work, and I'm fine with either one.

> @davehunt Which do you suggest?

I think we have to have a toggle method based on the disappointing situation that we can't have an expected initial state of the checkbox. Hopefully we can revisit this in the future. It would make sense to add a comment explaining the situation in the code.

rbillings commented:
I'm cleaning up and found this. There currently is comment code in there explaining what the code is doing - does it need additional commentary that we can't do it any other way? Does this test still need to be added at all?

davehunt commented:
> I'm cleaning up and found this. There currently is comment code in there explaining what the code is doing - does it need additional commentary that we can't do it any other way? Does this test still need to be added at all?

I've just reread the patch and the comments, and as far as I can tell there's still outstanding work to be done. I can't speak for whether this tests still needs to be added, but I have no reason to suspect it doesn't. It's probably worth seeing if the feature still exists, or asking @krupa. I would then suggest rebasing from upstream due to likely conflicts since branch this was last modified, and applying the suggested changes from the comments, and then pushing for review again.

rbillings commented:
I'll pass this on to someone else.
